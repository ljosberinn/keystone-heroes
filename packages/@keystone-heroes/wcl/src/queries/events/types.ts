export type BaseEvent<T extends Record<string, unknown>> = T & {
  timestamp: number;
  // pin: "0" but irrelevant
};

type EncounterStartEvent = BaseEvent<{
  type: "encounterstart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
  level: number;
  affixes: number[];
}>;

type CombatantInfoEvent = BaseEvent<{
  type: "combatantinfo";
  sourceID: number;
  gear: [];
  auras: {
    source: number;
    ability: number;
    stacks: number;
    icon: string;
    name?: string;
  }[];
  expansion: string;
  faction: number;
  specID: number;
  covenantID: number;
  soulbindID: number;
  strength: number;
  agility: number;
  stamina: number;
  intellect: number;
  dodge: number;
  parry: number;
  block: number;
  armor: number;
  critMelee: number;
  critRanged: number;
  critSpell: number;
  speed: number;
  leech: number;
  hasteMelee: number;
  hasteRanged: number;
  hasteSpell: number;
  avoidance: number;
  mastery: number;
  versatilityDamageDone: number;
  versatilityHealingDone: number;
  versatilityDamageReduction: number;
  talents: { id: number; icon: string }[];
  pvpTalents: { id: number; icon: string }[];
  artifact: { traitID: number; rank: number; spellID: number; icon: string }[];
  heartOfAzeroth: {
    traitID: number;
    rank: number;
    spellID: number;
    icon: string;
  }[];
}>;

export type ApplyDebuffEvent = BaseEvent<{
  type: "applydebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  targetInstance?: number;
  targetMarker?: number;
}>;

export type ApplyDebuffStackEvent = BaseEvent<{
  type: "applydebuffstack";
  stack: number;
  abilityGameID: number;
}>;

export type ApplyBuffEvent = BaseEvent<{
  type: "applybuff";
  sourceID: number;
  sourceInstance?: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
}>;

export type RemoveBuffEvent = BaseEvent<{
  type: "removebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type CastEvent = BaseEvent<{
  type: "cast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker?: number;
}>;

export type DamageEvent = BaseEvent<{
  type: "damage";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  mitigated?: number;
  unmitigatedAmount?: number;
  tick?: boolean;
  sourceMarker?: number;
  absorbed?: number;
  overkill?: number;
  buffs?: string;
}>;

export type BeginCastEvent = BaseEvent<{
  type: "begincast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

export type RemoveDebuffEvent = BaseEvent<{
  type: "removedebuff";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type SummonEvent = BaseEvent<{
  type: "summon";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type PhaseStartEvent = BaseEvent<{
  type: "phasestart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
}>;

export type HealEvent = BaseEvent<{
  type: "heal";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  sourceMarker?: number;
  targetMarker?: number;
  tick?: boolean;
  overheal?: number;
  absorbed?: number;
}>;

type EnergizeEvent = BaseEvent<{
  type: "energize";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  resourceChange: number;
  resourceChangeType: number;
  otherResourceChange: number;
  waste: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type ApplyBuffStackEvent = BaseEvent<{
  type: "applybuffstack";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  stack: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type InterruptEvent = BaseEvent<{
  type: "interrupt";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  sourceMarker?: number;
}>;

export type AbsorbEvent = BaseEvent<{
  type: "absorbed";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  attackerID: number;
  amount: number;
  extraAbilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type RefreshDebuffEvent = BaseEvent<{
  type: "refreshdebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

type DispelEvent = BaseEvent<{
  type: "dispel";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  isBuff: boolean;
  sourceMarker: number;
  targetMarker: number;
}>;

export type DeathEvent = BaseEvent<{
  type: "death";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  // only present if NPC kills player
  killerID?: number;
  killerInstance?: number;
  killingAbilityGameID?: number;
  targetMarker?: number;
}>;

export type AnyEvent =
  | DeathEvent
  | DispelEvent
  | RefreshDebuffEvent
  | AbsorbEvent
  | InterruptEvent
  | ApplyBuffStackEvent
  | EnergizeEvent
  | HealEvent
  | PhaseStartEvent
  | SummonEvent
  | RemoveDebuffEvent
  | BeginCastEvent
  | DamageEvent
  | CastEvent
  | RemoveBuffEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
  | ApplyDebuffEvent
  | CombatantInfoEvent
  | EncounterStartEvent;

export type AllTrackedEventTypes = (
  | CastEvent
  | DeathEvent
  | AbsorbEvent
  | DamageEvent
  | HealEvent
  | InterruptEvent
  | BeginCastEvent
  | ApplyBuffEvent
  | ApplyDebuffEvent
  | ApplyDebuffStackEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
)[];
