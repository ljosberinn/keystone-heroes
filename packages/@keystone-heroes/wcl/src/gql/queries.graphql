query InitialReportData($reportID: String!) {
  reportData {
    report(code: $reportID) {
      title
      startTime
      endTime

      region {
        slug
      }

      fights(translate: true, killType: Kills) {
        id
        startTime
        endTime

        keystoneLevel
        keystoneAffixes
        keystoneBonus
        keystoneTime

        averageItemLevel
        friendlyPlayers

        gameZone {
          id
        }

        maps {
          id
        }
      }
    }
  }
}

query EventData(
  $reportID: String!
  $startTime: Float!
  $endTime: Float!
  $hostilityType: HostilityType
  $sourceID: Int
  $dataType: EventDataType
  $abilityID: Float
  $targetID: Int
  $targetInstance: Int
  $limit: Int
  $filterExpression: String
) {
  reportData {
    report(code: $reportID) {
      events(
        startTime: $startTime
        endTime: $endTime
        hostilityType: $hostilityType
        sourceID: $sourceID
        dataType: $dataType
        abilityID: $abilityID
        targetID: $targetID
        targetInstanceID: $targetInstance
        limit: $limit
        filterExpression: $filterExpression
      ) {
        data
        nextPageTimestamp
      }
    }
  }
}

query PetActors($reportID: String!) {
  reportData {
    report(code: $reportID) {
      masterData {
        actors(type: "Pet") {
          gameID
          petOwner
          id
        }
      }
    }
  }
}

query EnemyNPCIds($reportID: String!, $fightIDs: [Int]!) {
  reportData {
    report(code: $reportID) {
      fights(killType: Kills, fightIDs: $fightIDs) {
        enemyNPCs {
          id
          gameID
        }
      }
    }
  }
}

query Table(
  $reportID: String!
  $fightIDs: [Int]!
  $startTime: Float!
  $endTime: Float!
) {
  reportData {
    report(code: $reportID) {
      table(startTime: $startTime, endTime: $endTime, fightIDs: $fightIDs)
    }
  }
}

query FightPulls($reportID: String!, $fightIDs: [Int]!) {
  reportData {
    report(code: $reportID) {
      fights(translate: true, killType: Kills, fightIDs: $fightIDs) {
        dungeonPulls {
          startTime
          endTime

          x
          y

          maps {
            id
          }

          boundingBox {
            minX
            maxX
            minY
            maxY
          }

          enemyNPCs {
            id
            gameID
            minimumInstanceID
            maximumInstanceID
          }
        }
      }
    }
  }
}
