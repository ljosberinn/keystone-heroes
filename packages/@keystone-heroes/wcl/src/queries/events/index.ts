import type { DungeonIDs } from "@keystone-heroes/db/data/dungeons";
import type { PlayableClass, Affixes } from "@keystone-heroes/db/types";

import {
  deathFilterExpression,
  filterEnemyDeathEvents,
  filterPlayerDeathEvents,
  filterProfessionEvents,
  filterRemarkableSpellEvents,
  remarkableSpellFilterExpression,
} from "../events/other";
import type {
  AllTrackedEventTypes,
  BeginCastEvent,
  DeathEvent,
} from "../events/types";
import { filterAffixEvents, getAffixExpression } from "./affixes";
import { getDungeonExpression, filterDungeonEvents } from "./dungeons";
import { recursiveGetEvents } from "./utils";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  return [
    deathFilterExpression,
    remarkableSpellFilterExpression,
    ...getDungeonExpression(dungeonID),
    ...getAffixExpression(affixes),
  ]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: PlayableClass }[]
): Promise<{
  allEvents: AllTrackedEventTypes[];
  playerDeathEvents: DeathEvent[];
  enemyDeathEvents: (DeathEvent | BeginCastEvent)[];
}> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  const allEvents = await recursiveGetEvents<AllTrackedEventTypes>({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });

  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    actorIDSet
  );
  const affixEvents = filterAffixEvents(allEvents, params.affixes);
  const professionEvents = filterProfessionEvents(allEvents);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    actorIDSet
    // remarkableSpellEvents
  );
  const enemyDeathEvents = filterEnemyDeathEvents(allEvents, actorIDSet);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
    enemyDeathEvents,
  };
};
