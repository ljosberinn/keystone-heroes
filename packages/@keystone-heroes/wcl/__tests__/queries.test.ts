import { DungeonIDs } from "@keystone-heroes/db/data";
import { Affixes, PlayableClass } from "@keystone-heroes/db/types";
import { graphql } from "msw";
import { setupServer } from "msw/node";

import { getEvents } from "../src/queries/events";
import allEvents from "./fixtures/allEvents.json";

describe("getEvents", () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  test("works", async () => {
    const startTime = 0;
    const endTime = 1;
    const reportID = "reportID";

    let filterExpression = "";

    server.use(
      graphql.query("EventData", (req, res, ctx) => {
        expect(req.body?.variables.startTime).toBe(startTime);
        expect(req.body?.variables.endTime).toBe(endTime);
        expect(req.body?.variables.reportID).toBe(reportID);
        expect(req.body?.variables.limit).toBe(10_000);

        filterExpression = req.body?.variables.filterExpression;

        return res(
          ctx.data({
            reportData: {
              report: {
                events: {
                  data: allEvents,
                  nextPageTimestamp: null,
                },
              },
            },
          })
        );
      })
    );

    const result = await getEvents(
      {
        affixes: [
          Affixes.Tormented,
          Affixes.Sanguine,
          Affixes.Volcanic,
          Affixes.Fortified,
        ],
        dungeonID: DungeonIDs.DE_OTHER_SIDE,
        startTime,
        endTime,
        reportID,
        fightID: 1,
      },
      [
        {
          actorID: 1,
          class: PlayableClass.DemonHunter,
        },
        {
          actorID: 2,
          class: PlayableClass.DemonHunter,
        },
        {
          actorID: 3,
          class: PlayableClass.DemonHunter,
        },
        {
          actorID: 5,
          class: PlayableClass.DemonHunter,
        },
        {
          actorID: 60,
          class: PlayableClass.DemonHunter,
        },
      ]
    );

    expect(result).toMatchSnapshot();
    expect(filterExpression).toMatchSnapshot();
  });
});
