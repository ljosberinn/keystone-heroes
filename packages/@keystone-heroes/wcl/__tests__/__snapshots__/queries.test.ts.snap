// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEvents works 1`] = `
Object {
  "allEvents": Array [
    Object {
      "abilityGameID": 209862,
      "absorbed": 7284,
      "amount": 36422,
      "fight": 3,
      "hitType": 3,
      "mitigated": 1108,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 619235,
      "type": "damage",
      "unmitigatedAmount": 8392,
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 5,
      "targetID": 35,
      "targetInstance": 1,
      "timestamp": 647832,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 209862,
      "amount": 39693,
      "fight": 3,
      "hitType": 1,
      "mitigated": 2688,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "timestamp": 945376,
      "type": "damage",
      "unmitigatedAmount": 10615,
    },
    Object {
      "abilityGameID": 209862,
      "amount": 80215,
      "fight": 3,
      "hitType": 1,
      "mitigated": 954,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "targetMarker": 2,
      "timestamp": 995245,
      "type": "damage",
      "unmitigatedAmount": 8424,
    },
    Object {
      "abilityGameID": 209862,
      "amount": 14272,
      "fight": 3,
      "hitType": 1,
      "mitigated": 856,
      "pin": "0",
      "sourceID": -1,
      "targetID": 4,
      "timestamp": 1019230,
      "type": "damage",
      "unmitigatedAmount": 8412,
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 1,
      "targetID": 16,
      "targetInstance": 2,
      "targetMarker": 8,
      "timestamp": 1286270,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 1,
      "targetID": 10,
      "targetInstance": 8,
      "timestamp": 1286270,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 1,
      "targetID": 17,
      "targetInstance": 7,
      "timestamp": 1286270,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 1,
      "targetID": 9,
      "targetInstance": 7,
      "timestamp": 1286270,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 1,
      "targetID": 9,
      "targetInstance": 5,
      "timestamp": 1286270,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 228626,
      "fight": 2,
      "pin": "0",
      "sourceID": 1,
      "targetID": 9,
      "targetInstance": 6,
      "timestamp": 1286270,
      "type": "applydebuff",
    },
    Object {
      "abilityGameID": 186257,
      "sourceID": 104,
      "targetID": -1,
      "timestamp": 2363666,
      "type": "cast",
    },
    Object {
      "abilityGameID": 204021,
      "sourceID": 5,
      "targetID": 96,
      "targetInstance": 5,
      "timestamp": 2370826,
      "type": "cast",
    },
    Object {
      "abilityGameID": 306830,
      "sourceID": 5,
      "targetID": -1,
      "timestamp": 2375912,
      "type": "cast",
    },
    Object {
      "abilityGameID": 323639,
      "sourceID": 105,
      "targetID": 96,
      "targetInstance": 3,
      "timestamp": 2378637,
      "type": "cast",
    },
    Object {
      "abilityGameID": 258925,
      "sourceID": 105,
      "targetID": -1,
      "timestamp": 2379268,
      "type": "cast",
    },
    Object {
      "abilityGameID": 200174,
      "sourceID": 103,
      "targetID": -1,
      "timestamp": 2379542,
      "type": "cast",
    },
    Object {
      "abilityGameID": 202138,
      "sourceID": 5,
      "targetID": -1,
      "timestamp": 2393784,
      "type": "cast",
    },
    Object {
      "abilityGameID": 260402,
      "sourceID": 104,
      "targetID": -1,
      "timestamp": 2394097,
      "type": "cast",
    },
    Object {
      "abilityGameID": 202137,
      "sourceID": 5,
      "targetID": -1,
      "timestamp": 2395049,
      "type": "cast",
    },
    Object {
      "abilityGameID": 212084,
      "sourceID": 5,
      "targetID": -1,
      "timestamp": 2397603,
      "type": "cast",
    },
    Object {
      "abilityGameID": 328231,
      "sourceID": 104,
      "targetID": -1,
      "timestamp": 2402325,
      "type": "cast",
    },
    Object {
      "abilityGameID": 288613,
      "sourceID": 104,
      "targetID": -1,
      "timestamp": 2404001,
      "type": "cast",
    },
    Object {
      "abilityGameID": 204021,
      "sourceID": 5,
      "targetID": 99,
      "targetInstance": 3,
      "timestamp": 2405723,
      "type": "cast",
    },
    Object {
      "abilityGameID": 207684,
      "sourceID": 5,
      "targetID": -1,
      "timestamp": 2412093,
      "type": "cast",
    },
    Object {
      "abilityGameID": 198589,
      "sourceID": 105,
      "targetID": -1,
      "timestamp": 2412874,
      "type": "cast",
    },
    Object {
      "abilityGameID": 345130,
      "sourceID": 5,
      "targetID": 103,
      "timestamp": 2826042,
      "type": "cast",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3186763,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3331087,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3424154,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3436647,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3489087,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3506311,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 60,
      "timestamp": 3512625,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3514282,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3517463,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3534137,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3559079,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3560518,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3563937,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3566890,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 60,
      "timestamp": 3568140,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3573319,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3576106,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3793055,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3889770,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3910966,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 4036935,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 4273729,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 4634252,
      "type": "death",
    },
    Object {
      "abilityGameID": 307195,
      "sourceID": 1859,
      "sourceMarker": 6,
      "targetID": 1859,
      "targetMarker": 6,
      "timestamp": 12294824,
      "type": "applybuff",
    },
    Object {
      "abilityGameID": 321422,
      "sourceID": 4,
      "targetID": 4,
      "timestamp": 12295325,
      "type": "applybuff",
    },
    Object {
      "abilityGameID": 321422,
      "sourceID": 1846,
      "sourceMarker": 2,
      "targetID": 1846,
      "targetMarker": 2,
      "timestamp": 12297167,
      "type": "applybuff",
    },
    Object {
      "abilityGameID": 226510,
      "amount": 2265944,
      "hitType": 1,
      "overheal": 83228,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1076,
      "targetInstance": 3,
      "tick": true,
      "timestamp": 12878155,
      "type": "heal",
    },
    Object {
      "abilityGameID": 226512,
      "amount": 10522,
      "hitType": 1,
      "mitigated": 165,
      "pin": "0",
      "sourceID": -1,
      "sourceMarker": 2,
      "targetID": 294,
      "targetMarker": 2,
      "tick": true,
      "timestamp": 12897350,
      "type": "damage",
      "unmitigatedAmount": 5426,
    },
    Object {
      "abilityGameID": 226512,
      "amount": 46138,
      "hitType": 1,
      "mitigated": 8212,
      "pin": "0",
      "sourceID": -1,
      "sourceMarker": 6,
      "targetID": 330,
      "targetMarker": 6,
      "tick": true,
      "timestamp": 13129280,
      "type": "damage",
      "unmitigatedAmount": 16354,
    },
    Object {
      "abilityGameID": 226512,
      "amount": 6006,
      "hitType": 1,
      "mitigated": 328,
      "pin": "0",
      "sourceID": -1,
      "targetID": 320,
      "tick": true,
      "timestamp": 13553089,
      "type": "damage",
      "unmitigatedAmount": 6334,
    },
    Object {
      "abilityGameID": 226512,
      "absorbed": 5417,
      "amount": 5417,
      "hitType": 3,
      "mitigated": 447,
      "pin": "0",
      "sourceID": -1,
      "targetID": 307,
      "tick": true,
      "timestamp": 14138169,
      "type": "damage",
      "unmitigatedAmount": 5864,
    },
  ],
  "playerDeathEvents": Array [
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3186763,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3331087,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3424154,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3436647,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3489087,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3506311,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 60,
      "timestamp": 3512625,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3514282,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3517463,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3534137,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3559079,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3560518,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3563937,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3566890,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 60,
      "timestamp": 3568140,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3573319,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 2,
      "timestamp": 3576106,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 3793055,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 3889770,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 3,
      "targetMarker": 7,
      "timestamp": 3910966,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 4036935,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 1,
      "timestamp": 4273729,
      "type": "death",
    },
    Object {
      "abilityGameID": 0,
      "pin": "0",
      "sourceID": -1,
      "targetID": 5,
      "timestamp": 4634252,
      "type": "death",
    },
  ],
}
`;

exports[`getEvents works 2`] = `"(target.type = \\"player\\" and type = \\"death\\") or (source.type = \\"player\\" and type = \\"cast\\" and ability.id IN (1719, 97462, 1161, 5246, 64382, 18499, 871, 107574, 12975, 46924, 184364, 167105, 262228, 227847, 307865, 324143, 325886, 317349, 118, 66, 80353, 45438, 55342, 190319, 108978, 12472, 84714, 235219, 205021, 12042, 198154, 314793, 307443, 324220, 314791, 114018, 6770, 2094, 2983, 37683, 31224, 1856, 137619, 79140, 185313, 121471, 277925, 13750, 51690, 323547, 323654, 328547, 328305, 192058, 2825, 32182, 198103, 108271, 8143, 98008, 79206, 114052, 207399, 108280, 16191, 198838, 114050, 191634, 108281, 192222, 192249, 198067, 51533, 114051, 320137, 58875, 324386, 320674, 326059, 328923, 10060, 73325, 19236, 8122, 81782, 319952, 200174, 47585, 34433, 15286, 228260, 47788, 64843, 64901, 200183, 2050, 265202, 34861, 88625, 33206, 246287, 109964, 62618, 47536, 123040, 325013, 323673, 324724, 327661, 186265, 186257, 61648, 109304, 131894, 19574, 193530, 321530, 201430, 19577, 186289, 266779, 260402, 288613, 328231, 308491, 324149, 325028, 322109, 115203, 119381, 169340, 122783, 132578, 122278, 115176, 123904, 137639, 122470, 310454, 326860, 325216, 327104, 20484, 22812, 1850, 106898, 102793, 29166, 132158, 197721, 102342, 740, 203651, 33891, 61336, 102558, 50334, 202770, 205636, 78675, 194223, 102560, 102543, 106951, 323764, 323546, 326434, 325727, 188501, 187827, 212084, 204021, 320341, 202137, 202138, 207684, 209258, 344867, 258925, 196718, 196555, 198589, 191427, 306830, 323639, 329554, 317009, 31884, 1022, 642, 6940, 853, 633, 498, 31850, 86659, 204018, 327193, 216331, 31821, 114158, 184662, 343527, 231895, 343721, 205191, 316958, 304971, 328204, 328278, 29893, 30283, 104773, 20707, 698, 342601, 333889, 113858, 1122, 267217, 111898, 265187, 267171, 113860, 205180, 321792, 325640, 312321, 325289, 48707, 48792, 61999, 49039, 327574, 46585, 51052, 212552, 48743, 194844, 219809, 221699, 55233, 108199, 49028, 49206, 207289, 42650, 63560, 275699, 152279, 47568, 279302, 207167, 51271, 312202, 315443, 311648, 324128)) or (type = \\"applydebuff\\" and target.type = \\"npc\\" and ability.id = 228626) or (ability.id = 357901 and type = \\"heal\\") or (type = \\"damage\\" and ability.id in (357865,357841,357708,357901,357525,356925,356923,355806,358784,358970,356667,356414,358894,355709,355737,358967)) or (ability.name = \\"Sanguine Ichor\\" and ((target.type = \\"player\\" and type = \\"damage\\") OR (target.type = \\"npc\\" and type = \\"heal\\"))) or (type = \\"damage\\" and target.type = \\"player\\" and ability.id = 209862)"`;
